lib/zlog.c

```diff
@@ -0,0 +1,702 @@
+#include "printfrr.h"
+#include "frrcu.h"
+#include "zlog.h"
+
+DEFINE_MTYPE_STATIC(LIB, LOG_MESSAGE,  "log message")
+DEFINE_MTYPE_STATIC(LIB, LOG_TLSBUF,   "log thread-local buffer")
+
+DEFINE_HOOK(zlog_init, (const char *progname, const char *protoname,
+			unsigned short instance, uid_t uid, gid_t gid),
+		       (progname, protoname, instance, uid, gid))
+DEFINE_KOOH(zlog_fini, (), ())
+DEFINE_HOOK(zlog_aux_init, (const char *prefix, int prio_min),
+			   (prefix, prio_min))
+
+char zlog_prefix[128];
+size_t zlog_prefixsz;
+int zlog_tmpdirfd = -1;
+
+/* these are kept around because logging is initialized (and directories
+ * & files created) before zprivs code switches to the FRR user;  therefore
+ * we need to chown() things so we don't get permission errors later when
+ * trying to delete things on shutdown
+ */
+static uid_t zlog_uid = -1;
+static gid_t zlog_gid = -1;
+
+DECLARE_ATOMLIST(zlog_targets, struct zlog_target, head);
+static struct zlog_targets_head zlog_targets;
+
+/* cf. zlog.h for additional comments on this struct.
+ *
+ * Note: you MUST NOT pass the format string + va_list to non-FRR format
+ * string functions (e.g. vsyslog, sd_journal_printv, ...) since FRR uses an
+ * extended prinf() with additional formats (%pI4 and the like).
+ *
+ * Also remember to use va_copy() on args.
+ */
+
+struct zlog_msg {
+	struct timespec ts;
+	int prio;
+
+	const char *fmt;
+	va_list args;
+
+	char *stackbuf;
+	size_t stackbufsz;
+	char *text;
+	size_t textlen;
+
+	/* This is always ISO8601 with sub-second precision 9 here, it's
+	 * converted for callers as needed.  ts_dot points to the "."
+	 * separating sub-seconds.  ts_zonetail is "Z" or "+00:00" for the
+	 * local time offset.
+	 *
+	 * Valid if ZLOG_TS_ISO8601 is set.
+	 * (0 if timestamp has not been formatted yet)
+	 */
+	uint32_t ts_flags;
+	char ts_str[32], *ts_dot, ts_zonetail[8];
+};
+
+/* thread-local log message buffering
+ *
+ * This is strictly optional and set up by calling zlog_tls_buffer_init()
+ * on a particular thread.
+ *
+ * If in use, this will create a temporary file in /var/tmp which is used as
+ * memory-mapped MAP_SHARED log message buffer.  The idea there is that buffer
+ * access doesn't require any syscalls, but in case of a crash the kernel
+ * knows to sync the memory back to disk.  This way the user can still get the
+ * last log messages if there were any left unwritten in the buffer.
+ *
+ * Sizing this dynamically isn't particularly useful, so here's an 8k buffer
+ * with a message limit of 64 messages.  Message metadata (e.g. priority,
+ * timestamp) aren't in the mmap region, so they're lost on crash, but we can
+ * live with that.
+ */
+
+#if defined(HAVE_OPENAT) && defined(HAVE_UNLINKAT)
+#define CAN_DO_TLS 1
+#endif
+
+#define TLS_LOG_BUF_SIZE	8192
+#define TLS_LOG_MAXMSG		64
+
+struct zlog_tls {
+	char *mmbuf;
+	size_t bufpos;
+
+	size_t nmsgs;
+	struct zlog_msg msgs[TLS_LOG_MAXMSG];
+	struct zlog_msg *msgp[TLS_LOG_MAXMSG];
```

Comment:
Yes, the 64 is an arbitrary limit, but... it only needs to be large enough to "diminish" the cost of the syscall to write out the messages.  I don't think there is much gain by increasing it further, but it can be tested by changing the `#define`.  And if it really helps, we should change it for everyone.  Not much point to make it configurable IMHO :).